packages/intl/src/MessageFormat/Parser/Parser.php

```diff
@@ -0,0 +1,627 @@
+    {
+        $name = $this->parseName();
+
+        if ($this->peek() === ':') {
+            $this->consumeChar(':');
+            $namespace = $name;
+            $name = $this->parseName();
+
+            return new Identifier($name, $namespace);
+        }
+
+        return new Identifier($name);
+    }
+
+    private function parseName(): string
+    {
+        $start = $this->peek();
+
+        if (! preg_match(self::NAME_START_REGEX, $start)) {
+            $this->throw("Invalid identifier start character: '{$start}'");
+        }
+
+        $buffer = $this->readChar();
+
+        while (! $this->isEof()) {
+            $char = $this->peek();
+
+            if (! preg_match(self::NAME_CHAR_REGEX, $char)) {
+                break;
+            }
+
+            $buffer .= $this->readChar();
+        }
+
+        return $buffer;
+    }
+
+    private function parseEscapedChar(array $escapable): string
+    {
+        $this->consumeChar('\\');
+        $char = $this->peek();
+
+        if (in_array($char, $escapable, true)) {
+            return $this->readChar();
+        }
+
+        $this->throw("Invalid escape sequence: \\{$char}.");
+    }
+
+    private function consumeKeyword(string $keyword): void
+    {
+        if ($this->peek(strlen($keyword)) !== $keyword) {
+            $this->throw("Expected keyword '{$keyword}'");
+        }
+
+        $this->pos += strlen($keyword);
+    }
+
+    private function consumeChar(string $expected): void
+    {
+        if ($this->isEof()) {
+            $this->throw("Expected `{$expected}` but reached end of input.");
+        }
+
+        $char = $this->readChar();
+
+        if ($char !== $expected) {
+            $this->pos--;
+            $this->throw("Expected `{$expected}` but found `{$char}`.");
+        }
+    }
+
+    private function consumeOptionalWhitespace(): void
+    {
+        $this->consumeWhitespace(false);
+    }
+
+    private function consumeRequiredWhitespace(): void
+    {
+        $this->consumeWhitespace(true);
+    }
+
+    private function consumeWhitespace(bool $required): void
+    {
+        $startPos = $this->pos;
+
+        while (! $this->isEof()) {
+            $char = $this->peek();
+
+            if (! preg_match('/^[\s\x{061C}\x{200E}\x{200F}\x{2066}-\x{2069}]/u', $char)) {
+                break;
+            }
+
+            $this->readChar();
+        }
+
+        if ($required && $this->pos === $startPos) {
+            $this->throw('Required whitespace not found.');
+        }
+    }
+
+    private function readChar(): string
+    {
+        if ($this->isEof()) {
+            return '';
+        }
+
+        $char = mb_substr($this->input, $this->pos, 1, 'UTF-8');
+        $this->pos++;
+
+        return $char;
+    }
+
+    private function peek(int $length = 1): string
+    {
+        if ($this->isEof()) {
+            return '';
+        }
+
+        return mb_substr($this->input, $this->pos, $length, 'UTF-8');
```

Comment:
Just FYI, I did a lot of performance testing with Tempest's view lexer, and stuff like `$this->isEof()` instead of directly writing `$this->pos >= $this->len` does have an impact when you're in a parsing context.

I don't think it would be too noticeable when parsing messages, since they are pretty small compared to view files. So I'm totally fine keeping it as is for now, but we might need to reconsider if this turns out to be a performance bottleneck after all.

Reply:
Right, this makes sense. I think this parser has some room for optimization in general, so if I need to debug performance, I'll look into that first 👍